### ISSUE TITLE
Property 'controlLabel' does not exist on type 'AbstractControl'. Due to using Ivy!

### ISSUE THREAD

**Issue by Pieter-1337** [angular/angular#36810]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/....
Not with certainty

### Is this a regression?

 Yes, the previous version in which this bug was not present was: ....
Yes: In Angular 8 this issue was not present

### Description
When building a form with Reactive Forms I decided to use an extension on FormControl to add a property "controlLabel" to my formControls, In Version 8 of Angular this works with no problem, however in Angular 9 when I try to read the extra property on my extended class It, the app will not compile and I will receive an error as followed:

ERROR in apps/reactive-forms/src/app/forms/form-with-state-matcher/form-with-state-matcher.component.html:2:12 - error TS2339: Property 'controlLabel' does not exist on type 'AbstractControl'.
    
  <div>{{form.controls['name'].controlLabel}}</div>

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-jg9tyw?file=src%2Fapp%2Fapp.component.ts

I Noticed here that turning off the "Use Ivy" flag in the StackBlitz settings tab solves the problem

## 🔥 Exception or Error
<pre><code>

Error in src/app/app.component.html (2:10)
Property 'controlLabel' does not exist on type 'AbstractControl'.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 9.1.0
Node: 10.13.0
OS: win32 x64

Angular: 9.1.1
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.0
@angular-devkit/build-angular     0.901.0
@angular-devkit/build-optimizer   0.901.0
@angular-devkit/build-webpack     0.901.0
@angular-devkit/core              9.1.0
@angular-devkit/schematics        9.1.0
@angular/.cdk.MODULES             <error>
@angular/cdk                      9.2.1
@angular/cli                      9.1.0
@angular/material                 9.2.1
@ngtools/webpack                  9.1.0
@schematics/angular               9.1.0
@schematics/update                0.901.0
rxjs                              6.5.5
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

**Comment by is2ei** [comment]
@Pieter-1337 

This does not work. But...

```html
<div [formGroup]="form">
  <div>{{form.controls['name'].controlLabel}}</div>
</div>
```


This works for me.

```html
<div [formGroup]="form">
  <div>{{form.controls.name['controlLabel']}}</div>
</div>
```

I'm not sure this is regression because this code occurs compiler error even in version 8.

```typescript
// app.component.ts
console.log(this.form.controls.name.controlLabel); // Compile error!!
```

In version 9, template type checking feature is available so it just check the type and warns it.

**Comment by Pieter-1337** [comment]
@is2ei It most definitely works in 8, using it as we speak for another project in the same manner.

Anyhow it I am asking my form control with a specific name for a property on it... if that form control is set as FormControlExtension in my case, it should be able to return all properties of that class, it does not make sense to assume that my object only contains the properties of the base class....

**Comment by is2ei** [comment]
@Pieter-1337 

> it should be able to return all properties of that class, it does not make sense to assume that my object only contains the properties of the base class....

If you don't want the compiler to check the type, you can use `any`.

```typescript
export class AppComponent implements OnInit  {
  form: any;

  constructor(private formBuilder: FormBuilder) {}

  ngOnInit() {
    this.form = this.formBuilder.group({
      name: new FormControlExtension(
        "",
        false,
        [Validators.required],
        "Testing"
      )
    });
  }
}
```

**Comment by Pieter-1337** [comment]
I think it should work with ivy just the way it did in older versions, what's the point of using a class if the compiler won't be able to check it's properties ... If i wanted to use "any" I would have used plain JS instead of TS ...

**Comment by alxhub** [comment]
Confirmed as reported.

This appears to be a template type-checking bug. The error itself is legitimate (it's illegal to access `.controlLabel` on a value of type `AbstractControl`, but it was not caught by View Engine in either type-checking mode (basic or `fullTemplateTypeCheck`), so per our type-checking compatibility story, Ivy should not be surfacing the error unless in `strictTemplates` mode. This will require some digging to understand more completely.

**Comment by JoostK** [comment]
This is working properly as far as I can see. I get the exact same error in Angular 8 as I get in Angular 9/10, regardless of whether View Engine or Ivy is used.

I think the difference can be found in the `aot` flag, that was switched to `true` when upgrading to Angular 9. I AOT mode the compiler performs these type checks, which it would not in JIT mode. When AOT is enabled, there's currently no flag to fully disable type checking (as would be the case in JIT).

In conclusion, I don't think there is a bug here.  This change is documented [in the v9 upgrade instructions](https://v9.angular.io/guide/updating-to-version-9):

> CLI apps compile in AOT mode by default (which includes template type-checking). Users who only built with JIT before may see new type errors. See our template type-checking guide for more information and debugging tips.

**Comment by samvloeberghs** [comment]
@joost-de-vries can you please consider rechecking this and re-add the label `state:confirmed`?
Please check this reproduction where the bug is happening:

https://github.com/samvloeberghs/strict-templates-reactive-forms-bug

Steps to reproduce:

1. `npm i`
2. `npm run build -- --prod`
3. issues error: 

<details>
   <summary>Show error messages</summary>

```
ERROR in src/app/app.component.html:2:28 - error TS2739: Type 'AbstractControl' is missing the following properties from type 'FormControl': registerOnChange, registerOnDisabledChange, _applyFormState

2   Name: <input type="text" [formControl]="form.controls['name']" />
                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  src/app/app.component.ts:6:16
    6   templateUrl: './app.component.html',
                     ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.
src/app/app.component.html:3:33 - error TS2322: Type 'AbstractControl' is not assignable to type 'FormControl'.

3   Firstname: <input type="text" [formControl]="form.controls['firstName']" />
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  src/app/app.component.ts:6:16
    6   templateUrl: './app.component.html',
                     ~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component AppComponent.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! strict-templates-reactive-forms-bug@0.0.0 build: `ng build "--prod"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the strict-templates-reactive-forms-bug@0.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
```
</details>

I tried several approaches targeting the right `formControl`:
1. using `[formControl]="form.controls['name']"`
2. using `[formControl]="form.get('firstName')"`
3. using a public property on the component itself, and attaching it to the formGroup `[formControl]="middleName"`
 (not really liking this approach as it exposes the formControl itself while it should potentially only be used within the formGroup)

```html
<!-- app.component.html -->
<form [formGroup]="form">
  Name: <input type="text" [formControl]="form.controls['name']" />
  Firstname: <input type="text" [formControl]="form.get('firstName')" />
  Middle name: <input type="text" [formControl]="middleName" />
</form>
```

```ts
// app.component.ts
import { Component } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  title = 'strict-templates-reactive-forms-bug';
  middleName = new FormControl('');
  form = new FormGroup({
    name: new FormControl(''),
    firstName:  new FormControl(''),
    middleName: this.middleName
  });
}
```



Issue starter uses an extension of FormControl, but it's actually also happening with a normal FormControl.

**Comment by JoostK** [comment]
@samvloeberghs I think you meant to address me (although I do know @joost-de-vries and we live in the same city, I don't think he'll respond :p)

The issue as reported here is about template type checking errors that occur in combination with Forms APIs, particularly accessing the property `controlLabel` from a supposed `FormControl`. This supposedly did not report an error in earlier versions of Angular, however it has always detected this particular kind of _unsafe_ access provided that AOT compilation is used (template type checking is only done in AOT). So from the compiler's perspective, there is no bug with the template type checking capability.

The examples you bring up are similar, but the scenario is a bit different given that `strictTemplates` is enabled. This enables more accurate type checking and as you've found out, the Forms API doesn't really play nicely under these circumstances. The return type of the index signature is `AbstractControl`, so your knowledge of it actually being a concrete `FormControl` is not reflected in the Forms's APIs type design, so the template type checker doesn't have this knowledge. There's issues like #36405, #17000, #13721 that request APIs to address this limitation in one way or another.

So although the examples you show are showing errors that may be considered undesirable (I fully agree with you here), they are the result of limitations in the Forms APIs, not limitations or bugs in the type checking machinery. This is why I revoked the confirmed label, as this is not actionable from the compiler's perspective.

**Comment by samvloeberghs** [comment]
> @samvloeberghs I think you meant to address me (although I do know @joost-de-vries and we live in the same city, I don't think he'll respond :p)
> 
> The issue as reported here is about template type checking errors that occur in combination with Forms APIs, particularly accessing the property `controlLabel` from a supposed `FormControl`. This supposedly did not report an error in earlier versions of Angular, however it has always detected this particular kind of _unsafe_ access provided that AOT compilation is used (template type checking is only done in AOT). So from the compiler's perspective, there is no bug with the template type checking capability.
> 
> The examples you bring up are similar, but the scenario is a bit different given that `strictTemplates` is enabled. This enables more accurate type checking and as you've found out, the Forms API doesn't really play nicely under these circumstances. The return type of the index signature is `AbstractControl`, so your knowledge of it actually being a concrete `FormControl` is not reflected in the Forms's APIs type design, so the template type checker doesn't have this knowledge. There's issues like #36405, #17000, #13721 that request APIs to address this limitation in one way or another.
> 
> So although the examples you show are showing errors that may be considered undesirable (I fully agree with you here), they are the result of limitations in the Forms APIs, not limitations or bugs in the type checking machinery. This is why I revoked the confirmed label, as this is not actionable from the compiler's perspective.

Thank you very much for your clear answer. I will
have to find another workaround or use the one described with the FormControl properties directly on the instance of the component.

**Comment by JoostK** [comment]
@samvloeberghs There's various options available in your case. You could disable `strictInputTypes` and leave the rest of strict checking enabled, but this will have an adverse effect on type checking in general (all bindings won't be checked in that case, nor will type inference across directives such as `ngFor` work, so it's quite limiting to have that disabled). An alternative would be to wrap the expressions in `$any` "calls", which are `as any` casts but within an Angular expression. This requires that the template expressions are changed, but avoids having to declare all form controls on the component instance.

**Comment by JoostK** [comment]
I'll go ahead and close this as not a compiler issue/a regression from earlier versions. The difference in behavior is due to AOT being enabled by default, which is documented in the v9 release notes.

Improvements to forms API to better behave with type checking is found in issues like #36405, #17000 and #13721.
